backends/ebpf/runtime/ebpf_kernel.h

```diff
@@ -23,30 +23,48 @@ limitations under the License.
 #ifndef BACKENDS_EBPF_RUNTIME_EBPF_KERNEL_H_
 #define BACKENDS_EBPF_RUNTIME_EBPF_KERNEL_H_
 
-#include <linux/pkt_cls.h>  // TC_ACT_OK, TC_ACT_SHOT
-#include <string.h>         // memset()
-#include <unistd.h>         // syscall()
-#include <sys/syscall.h>    // __NR_bpf
 #include "ebpf_common.h"
 
 
+/* If we operate in user space we only need to include bpf.h and
+ * define the userspace API macros.
+ * For kernel programs we need to specify a list of kernel helpers. These are
+ * taken from here: https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/bpf_helpers.h
+ */
+#ifdef CONTROL_PLANE // BEGIN EBPF USER SPACE DEFINITIONS
+#include "bpf.h" // bpf_obj_get/pin, bpf_map_update_elem
+
+#define BPF_USER_MAP_UPDATE_ELEM(index, key, value, flags)\
+    bpf_map_update_elem(index, key, value, flags)
+#define BPF_OBJ_PIN(table, name) bpf_obj_pin(table, name)
+#define BPF_OBJ_GET(name) bpf_obj_get(name)
+
+#else // BEGIN EBPF KERNEL DEFINITIONS
+#include <linux/pkt_cls.h>  // TC_ACT_OK, TC_ACT_SHOT
+#include "linux/bpf.h"  // types, and general bpf definitions
 
-/* Additional headers */
-# define printk(fmt, ...)                                               \
+#define bpf_printk(fmt, ...)                                            \
                 ({                                                      \
                         char ____fmt[] = fmt;                           \
                         bpf_trace_printk(____fmt, sizeof(____fmt),      \
                                      ##__VA_ARGS__);                    \
                 })
 
 /* helper functions called from eBPF programs written in C */
-static void *(*bpf_map_lookup_elem)(void *map, void *key) =
+static void *(*bpf_map_lookup_elem)(void *map, const void *key) =
```

Comment:
I was hoping we would not need this junk anymore.
Isn't there a header with this stuff in libbpf?

Reply:
I will see if I can strip more later. Last time I checked those were still necessary.